Fixes several errors on input and output parameters. (#286)
cfrg/draft-irtf-cfrg-opaque: draft-irtf-cfrg-opaque.md

- DeriveKeyPair(seed): Derive a private and public key pair deterministically
  from a seed.
- Noe: The size of a serialized OPRF group element.

> - Nok: The size of an OPRF private key.

Author: I think it's not used, and/or its definition is not correct.

> - Nok: The size of an OPRF private key.

Reviewer: It is used in `CreateCredentialResponse()`, 1st step, to derive the per-user OPRF key.

> The public input `info` is currently set to the empty string.

Reviewer 2: I totally don't care about the wording here, so both seem fine to me.

A Memory Hard Function (MHF) is a slow and expensive cryptographic hash function
with the following API:

- Harden(msg): Apply a memory-hard function to strengthen the input `msg`

>   against offline dictionary attacks.

Reviewer: What do you suggest as parameters? Should we refer to the default parameters of the MHF?

material for the AKE from the Envelope. This function raises an error if the
private seed cannot be used for recovering authentication material from the
input envelope.

> - Ne: The size of the envelope, which may vary between mechanisms.

Reviewer 2: ```suggestion

The size of the envelope is a constant denoted as `Ne`. Its size may vary between key recovery methods.
```

Author: I used the convention in the previous section where constants are explicitly itemized. I suggest this change because finding this parameter in the middle of a sentence is more difficult to parse.

Reviewer 2: Well, it's not a function, so the sentence introducing the list would need to change if we were to take this suggestion.

Author: updated the sentence.

Reviewer 2: Looks good!

# Protocol Overview {#protocol-overview}

OPAQUE consists of two stages: registration and online authentication.
In the first stage, a client uses its password to generate authentication
material that server stores in a credential file.
In the second stage the client recovers its authentication material from server

> and uses it to perform a mutually authenticated key exchange.
> 
> ## Setup
> Previously to both stages, client and server agree on a configuration, which

Reviewer 2: ```suggestion
and uses it to perform a mutually authenticated key exchange. For both stages, client and 
server agree on a configuration, which
```

Author: I feel a lack of a section called Setup, like any other protocols has.

Reviewer 2: I don't think it's necessary.

Author: I think it is necessary. Shall we move this discussion to the list?

Reviewer 2: No, I don't think so. This is an editorial consideration.

Author: Reverted and will be discussed in #289

fully specifies the cryptographic algorithm dependencies necessary to run the
protocol; see {{configurations}} for details.

> Before the registration phase, the client chooses its password, and the server
> chooses its own pair of keys (server_private_key, server_public_key) for use
> with the AKE, and chooses an oprf_seed of `Nseed` bytes. The server can use the
> same pair of keys with multiple clients and can opt to use multiple seeds (so
> long as they are kept consistent for each client).

Reviewer 2: ```suggestion
```
This is redundant with text below in the registration overview.

registration that it stores in a credential file alongside other client
registrations as needed.

The registration flow, shown below, uses three messsage structures named
`RegistrationRequest`, `RegistrationResponse`, and `RegistrationRecord`. Their
contents and wire format are defined in {{registration-messages}}.

~~~
    creds                                   parameters
      |                                         |
      v                                         v
    Client                                    Server
    ------------------------------------------------

>                       request

Reviewer 2: ```suggestion
                      RegistrationRequest
```

Author: I changed this twice, I still didn't understand what is the convention.

What must be on top of the arrow? the object type or an instance of such an object. I can change either way, but want to know what is the convention to follow.

Reviewer 2: I don't know if we have a convention, but using the type here seems most clear.

             ------------------------->

>                       response

Reviewer 2: ```suggestion
                      RegistrationResponse
```

             <-------------------------

>                       record

Reviewer 2: ```suggestion
                      RegistrationRecord
```

corresponding to the client. The client outputs two values, an `export_key`
(matching that from registration) and a `session_key`, the latter of which
is the primary AKE output. The server outputs a single value `session_key`
that matches that of the client. Upon completion, the client and server can
use these values as needed.

The online authentication flow is shown below:

~~~
    creds                             (parameters, record)
      |                                         |
      v                                         v
    Client                                    Server
    ------------------------------------------------

>                         ke1

Reviewer 2: ```suggestion
                        KE1
```

             ------------------------->

>                         ke2

Reviewer 2: ```suggestion
                        KE2
```

             <-------------------------

>                         ke3

Reviewer 2: ```suggestion
                        KE3
```

             ------------------------->
   ------------------------------------------------
      |                                         |
      v                                         v
(export_key, session_key)                  session_key
~~~

> These messages correspond, respectively, to structures named `KE1`, `KE2`,
> and `KE3` that carry information of the concurrent execution of the
> credential retrieval process and the authenticated key exchange:

Reviewer 2: ```suggestion
These messages carry information of the concurrent execution of the
credential retrieval process and the authenticated key exchange:
```

Input:
- server_public_key, The encoded server public key for the AKE protocol.
- client_public_key, The encoded client public key for the AKE protocol.

> - server_identity, The optional server identity.

Reviewer 2: ```suggestion
- server_identity, The optional server identity, an octet string.
```

> - client_identity, The optional client identity.

Reviewer 2: ```suggestion
- client_identity, The optional client identity, an octet string.
```

Input:
- randomized_pwd, randomized password.

> - server_public_key, The encoded server public key for the AKE protocol.

Reviewer 2: ```suggestion
- server_public_key, The encoded server public key for
the AKE protocol, an octet string.
```

> - server_identity, The optional server identity.

Reviewer 2: ```suggestion
- server_identity, The optional server identity, an octet string.
```

> - client_identity, The optional client identity.

Reviewer 2: ```suggestion
- client_identity, The optional client identity, an octet string.
```

- randomized_pwd, randomized password.
- server_public_key, The encoded server public key for the AKE protocol.
- envelope, the client's `Envelope` structure.

> - server_identity, The optional server identity.

Reviewer 2: ```suggestion
- server_identity, The optional server identity, an octet string.
```

> - client_identity, The optional client identity.

Reviewer 2: ```suggestion
- client_identity, The optional client identity, an octet string.
```

- password, an opaque byte string containing the client's password.
- blind, an OPRF scalar value.
- response, a RegistrationResponse structure.

> - server_identity, the optional server identity.

Reviewer 2: ```suggestion
- server_identity, the optional server identity, an octet string.
```

> - client_identity, the optional client identity.

Reviewer 2: ```suggestion
- client_identity, the optional client identity, an octet string.
```

Author: Does this applies to all occurrences of `*_identity`?

Reviewer 2: When used as an octet string, yes.

The server stores the `record` object as the credential file for each client
along with the associated `credential_identifier` and `client_identity` (if
different). Note that the values `oprf_seed` and `server_private_key` from the
server's setup phase must persist for the online authentication phase. The
`oprf_seed` value SHOULD be used
for all clients; see {{preventing-client-enumeration}}. The `server_private_key`
may be unique for each client.

> # Online Authentication {#online-phase}

Reviewer 2: ```suggestion
# Online Authenticated Key Exchange {#online-phase}
```
One output is a session key, after all.

Author: My understanding of the draft is that during Online Authentication two tasks happen, and one of them is an Authenticated Key Exchange (AKE). However, Online Authentication, I think is a more general title for describing this big phase, later there is one section only of AKE.

Reviewer 2: OPAQUE is fundamentally a key exchange protocol, so the existing seems more accurate.

Author: I am talking about the `Online Authentication phase`, which is a combination of several primitives, one of them is AKE (which seems to be the most important, but not the only one).

Author: After talked offline with Chris Wood, we agreed the current version is ok. So, I'm reverting.

The generic outline of OPAQUE with a 3-message AKE protocol includes three messages
ke1, ke2, and ke3, where ke1 and ke2 include key exchange shares, e.g., DH values, sent
by the client and server, respectively, and ke3 provides explicit client authentication and
full forward security (without it, forward secrecy is only achieved against eavesdroppers,
which is insufficient for OPAQUE security).

> This section describes the online authentication protocol flow,

Reviewer 2: ```suggestion
This section describes the online authenticated key exchange protocol flow,
```

                         ke2
              <-------------------------

    (ke3,
    session_key,
    export_key) = ClientFinish(client_identity, password,
                              server_identity, ke2)
  (ke3,
  session_key,
  export_key) = ClientFinish(client_identity,
                             server_identity,
                             ke1, ke2)

                         ke3
              ------------------------->

                       session_key = ServerFinish(ke3)
~~~

Both client and server may use implicit internal state objects to keep necessary
material for the OPRF and AKE, `client_state` and `server_state`, respectively.
Both client and server may use implicit internal state objects, `client_state`
and `server_state`, to keep necessary material for the OPRF and AKE.

The client state may have the following named fields:

- password, the input password; and
- blind, the random blinding scalar returned by `Blind()`, of length Nok; and
- blind, the random blinding scalar returned by `Blind()`; and

> - client_ake_state, the client's AKE state if necessary.

Author: fixed.

                         ke2
              <-------------------------

  (ke3,
  session_key,
  export_key) = ClientFinish(client_identity,
                             server_identity,
                             ke1, ke2)

                         ke3
              ------------------------->

                       session_key = ServerFinish(ke3)
~~~

Both client and server may use implicit internal state objects, `client_state`
and `server_state`, to keep necessary material for the OPRF and AKE.

The client state may have the following named fields:

- password, the input password; and

> - blind, the random blinding scalar returned by `Blind()`; and

Author: This parameter was not used

- password, an opaque byte string containing the client's password.

Output:
- ke1, a KE1 structure.

Steps:
1. request, blind = CreateCredentialRequest(password)
2. state.password = password
3. state.blind = blind
4. auth_init = Start()
5. Create KE1 ke1 with (request, auth_init)
~~~

~~~
ClientFinish(client_identity, server_identity, ke1, ke2)

> 

Author: I fixed this, please double-check. I may missed something.

State:
- state, a ClientState structure

> Input:

Author: yes, I just update the pr to do that.

           record, credential_identifier, oprf_seed, ke1, client_identity)

Input:
- server_identity, the optional server identity.
- server_private_key, the server's private key for the AKE protocol.
- server_public_key, the server's public key for the AKE protocol.
- record, the client's RegistrationRecord structure.
- credential_identifier, an identifier that uniquely represents the credential.
- oprf_seed, the server's OPRF seed.
- ke1, a KE1 message structure.
- client_identity, the optional client identity.

Output:

> - ke2, a KE2 structure.

Reviewer 2: Should this output type change?

Author: not sure, change to what type? (didn't get your question, could you please elaborate)

Reviewer 2: Well, the function returns something of type `Response`, not `KE2`.

Author: No. It returns the result of calling `Response` function. (https://github.com/cfrg/draft-irtf-cfrg-opaque/blob/515717e8b6f19d427d6190f99450a850cfb709aa/draft-irtf-cfrg-opaque.md?plain=1#L1496)

Reviewer 2: Oh, right! Sheesh. We need to really fix the names in this spec. 🤦

~~~

### 3DH Client Functions {#ake-client}

~~~
Start()

State:
- state, the AKE client state.

Output:
- auth_init, a AuthInit structure.

Steps:
1. client_nonce = random(Nn)
2. client_secret, client_keyshare = GenerateAuthKeyPair()
3. Create AuthInit auth_init with (client_nonce, client_keyshare)
4. state.client_secret = client_secret
5. Output auth_init
~~~

~~~
ClientFinalize(client_identity, client_private_key, server_identity,
               server_public_key, ke1, ke2)

State:
- state, the AKE client state.

Input:
- client_identity, the optional client identity.
- client_private_key, the client's private key for the AKE protocol.
- server_identity, the optional server identity.
- server_public_key, the server's public key for the AKE protocol.
- ke1, a KE1 structure.
- ke2, a KE2 structure.

Output:
- auth_finish, an AuthFinish structure.
- session_key, the shared session secret.

Exceptions:
- HandshakeError, when the handshake fails.

Steps:
1. ikm = TripleDHIKM(state.client_secret, ke2.auth_response.server_keyshare,
                     state.client_secret, server_public_key,
                     client_private_key, ke2.auth_response.server_keyshare)

> 2. preamble = Preamble(client_identity, ke1, server_identity, ke2)

Reviewer 2: ke1 was previously in the state. Why change that here?

Author: Because the `AKE client state` only carries values from the AKE.
The external function (the one that calls the AKE) is the one that knows how to build KE1 during Start().

Reviewer 2: This seems like an aesthetic change. Can we please keep these separate?

Author: Not sure how to implement your suggestion.

Just to add more context and information about the fix:

There are two states, one wrapping the other: 
A: The Main State: (https://github.com/cfrg/draft-irtf-cfrg-opaque/blob/515717e8b6f19d427d6190f99450a850cfb709aa/draft-irtf-cfrg-opaque.md?plain=1#L988)
which contains:
- password, the input password; and
- blind, the random blinding scalar returned by `Blind()`; and
- B: client_ake_state, the client's AKE state if necessary.
 
and B: the client_ake_state (https://github.com/cfrg/draft-irtf-cfrg-opaque/blob/515717e8b6f19d427d6190f99450a850cfb709aa/draft-irtf-cfrg-opaque.md?plain=1#L1249)
containing:
- client_secret, an opaque byte string of length Nsk.

---

ke1 must live in the state, yes. the question is in which one?

We cannot store ke1 in state B, because ke1 is constructed as (https://github.com/cfrg/draft-irtf-cfrg-opaque/blob/515717e8b6f19d427d6190f99450a850cfb709aa/draft-irtf-cfrg-opaque.md?plain=1#L570
)
```
- `KE1` is composed of the `CredentialRequest` and `AuthInit` messages.
```

AuthInit is created by the AKE, but CredentialRequest is not.
So the AKE cannot construct KE1, this is why KE1 must be passed as a parameter.

Hence, ke1 must live in the main state A. -- This part needed to be updated, and this is fixed now.

Reviewer 2: Well, this PR changes state from "generic state" to "AKE client state," and I'm asking that we revert that change and address it separately.
